apps/prairielearn/src/ee/lib/lti13.ts

```diff
@@ -211,3 +258,227 @@ export async function validateLti13CourseInstance(
+      });
+    }
+
+    const output = await queryRow(
+      sql.sync_lti13_lineitems,
+      {
+        lti13_course_instance_id: instance.lti13_course_instance.id,
+      },
+      z.object({
+        updated: z.string(),
+        deleted: z.string(),
+      }),
+    );
+
+    job.info(
+      `\nSummary of PrairieLearn changes: ${output.updated} updated, ${output.deleted} deleted.`,
+    );
+    job.info('Done.');
+  });
+}
+
+export async function create_and_link_lineitem(
+  instance: any,
+  job: ServerJob,
+  assessment: {
+    label: string;
+    id: string;
+    url: string;
+  },
+) {
+  const createBody: Lti13Lineitem = {
+    id: 'new_lineitem', // will be ignored/overwritten by the LMS platform
+    scoreMaximum: 100,
+    label: assessment.label,
+    resourceId: assessment.id,
+    // tag ???
+    // startDateTime null
+    // endDateTime null
+    'https://canvas.instructure.com/lti/submission_type': {
+      type: 'external_tool',
+      external_tool_url: assessment.url,
+    },
+  };
+
+  job.info(
+    `Creating assignment for ${assessment.label} in ${instance.lti13_instance.name} ${instance.lti13_course_instance.context_label}`,
+  );
+
+  const token = await access_token(instance.lti13_instance.id);
+  const item: Lti13Lineitem = await fetchRetry(instance.lti13_course_instance.lineitems, {
+    method: 'POST',
+    headers: {
+      Authorization: `Bearer ${token}`,
+      'Content-type': 'application/vnd.ims.lis.v2.lineitem+json',
+    },
+    body: JSON.stringify(createBody),
+  });
+
+  job.info('Associating PrairieLearn assessment with the new assignment');
+
+  await link_assessment(instance.lti13_course_instance.id, assessment.id, item);
+  job.info('Done.');
+}
+
+export async function query_and_link_lineitem(
+  instance: any,
+  lineitem_id: string,
+  assessment_id: string | number,
+) {
+  const lineitems = await get_lineitems(instance);
+  const item = lineitems.find(({ id }) => id === lineitem_id);
+
+  if (item) {
+    await link_assessment(instance.lti13_course_instance.id, assessment_id, item);
+  } else {
+    throw new HttpStatusError(400, 'Lineitem not found');
+  }
+}
+
+export async function unlink_assessment(
+  lti13_course_instance_id: string,
+  assessment_id: string | number,
+) {
+  await queryAsync(sql.delete_lineitem_by_assessment_id, {
+    lti13_course_instance_id,
+    assessment_id,
+  });
+}
+
+export async function link_assessment(
+  lti13_course_instance_id: string,
+  assessment_id: string | number,
+  lineitem: Lti13Lineitem,
+) {
+  await queryAsync(sql.update_lineitem, {
+    lti13_course_instance_id,
+    lineitem_id: lineitem.id,
+    lineitem: JSON.stringify(lineitem),
+    assessment_id,
+  });
+}
+
+function sleep(delay: number) {
+  return new Promise((resolve) => setTimeout(resolve, delay));
+}
+
+/* Throttling notes
+// https://canvas.instructure.com/doc/api/file.throttling.html
+// 403 Forbidden (Rate Limit Exceeded)
+// X-Request-Cost
+// X-Rate-Limit-Remaining
+*/
```

Comment:
What are these notes relevant to? I don't see any special handling of either the status code or headers.

Reply:
`response.ok` evaluates to true if 200 <= response status < 300. The current version of the code doesn't trigger off of 403 differently than any other 400 or 500 error (it retries), but those notes were the beginning of that logic.